qiskit/circuit/quantumcircuit.py

```diff
@@ -491,6 +494,34 @@ def __str__(self) -> str:
         return str(self.draw(output="text"))
 
     def __eq__(self, other) -> bool:
+        """Check if two circuits are equal.
+        This will check :attr:`~QuantumCircuit.calibrations` and :attr:`~QuantumCircuit.global_phase`
+        and will not consider :attr:`~QuantumCircuit.data`, :attr:`~QuantumCircuit.name`,
+        :attr:`~QuantumCircuit.metadata` when making determination.
+        The reason for this is due to being hard and computationally expensive to perform.
+        Please refer to :class:`.Operator` if you need to achieve that level of comparison.
```

Comment:
The equality does take into account `QuantumCircuit.data` -- it doesn't require the insertion order of the gates to be the same, since the circuit implements the same unitary still 🙂 The actual comparison is implemented on the level of the DAG representation, you can check the implementation here; https://github.com/Qiskit/qiskit/blob/4c1de131201164d5a2116342b9a431b34a92d917/crates/circuit/src/dag_circuit.rs#L2404.

If you still would like to contribute, we should update this to state the the DAG representation of both circuits, as well as the qubits, registers and metadata is compared. I like your note about the operator, we could add this stating that "to compare whether circuits with different gate representations implement the same unitary, refer to the Operator class".